parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/state/ShardManager.java

```diff
@@ -298,4 +321,18 @@ private void initMetrics() {
         numberOfShardsGauge = pcMetrics.gaugeFromMetricDef(PCMetricsDef.NUMBER_OF_SHARDS,
                 this, shardManager -> shardManager.processingShards.keySet().size());
     }
+
+    // get expired items count from retryQueue
+    private long getExpiredItemCntFromRetryQueue() {
+        long count = 0;
+        for (WorkContainer<?, ?> workContainer : retryQueue) {
+            if (workContainer.getDelayUntilRetryDue().isNegative() || workContainer.getDelayUntilRetryDue().isZero()) {
+                count ++;
```

Comment:
should do another check before incrementing count as WorkContainer can be still in retryQueue but at the same time already picked for processing - so should be using either `workContainer.isAvailableToTakeAsWork()` or 
`isNotInFlight() && !isUserFunctionSucceeded()` as we already verified that delay passed by this point.
so basically it would be 
```
// count work that is not waiting for retry backoff delay to pass and is not inflight or succeeded already
if(workContainer.isDelayPassed()){
    if(workContainer.isNotInFlight() && !workContainer.isUserFunctionSucceeded()){
        count++;
    }
} else {
     // early stop since retryQueue is sorted by retryDueAt
     break;
}
```
The remainder of the queue - those that have their delay not passed yet are OK in that regard as it cannot be inflight if its still waiting for retry delay to pass.

Reply:
fixed

Reply:
`// 2. the container has been selected and it is inflight but we already slashed them from availableWorkContainerCnt, so should be counted in`
This is not correct - we do not count inflight containers as available for selection.
The intent of the `getCountOfWorkAwaitingSelection` is to get a number for all work that is ready to be processed - to determine if Consumer should poll for more messages - or there is already enough queued for processing - so that excludes work that is already being processed / inflight.

In original code - the condition was
```
public boolean isAvailableToTakeAsWork() {
        return isNotInFlight() && !isUserFunctionSucceeded() && isDelayPassed();
    }
```
And the updated logic should really reflect the original logic.